codice/ddf
# [2.13.x] DDF-04469 Added Online Certificate Status Protocol (OCSP) support

#### What does this PR do?
Adds Online Certificate Status Protocol (OCSP) support to DDF. An OCSP server can be queried to verify whether a certificate is revoked or not. If an error occurs during the communication with the OCSP server. An alert will be posted to the admin console.

#### Who is reviewing it? 
@bakejeyner 

#### Select relevant component teams: 
@codice/security 

#### Ask 2 committers to review/merge the PR and tag them here.
@brjeter
@garrettfreibott
@ricklarsen - Documentation
@stustison

#### How should this be tested?
**THESE TESTING STEPS ARE FOR AN HTTP OCSP SERVER. I WILL UPDATE THEM WITH AN HTTPS OCSP SERVER STEPS.**

1. Create a p12 file that can be used in Chrome using DDF's demo CA
  * `cd <DDF_HOME>/etc/certs/demoCA`
  * Create a .p12 file from the ca cert and ca key
    * `openssl pkcs12 -export -out ca.p12 -inkey private/cakey.pem -in cacert.pem`
  * Open the KeyStoreExplorer application and import the `ca.p12` file above
    * `Create a new KeyStore`
    * Select `PKCS #12`
    * Go to `Tools > Import Key Pair > PKS#12` and import the `ca.p12` file
    * Right click on the newly imported keypair and select `Sign > Sign New Key Pair`
      * Add `CN=<hostName>` to `Name`
    * Right click on the newly created keypair and select `Export > Export Key Pair`

2. Import that exported keypair into Google Chrome.
  * `Settings > Advanced > Manage Certificates`
  * On Mac:
    * `File > Import Items > Login > <Exported-KeyPair>.p12`

3. Create an OCSP server
  * `openssl req -new -nodes -out ocspSigning.csr -keyout ocspSigning.key`
    * Make sure to set the CommonName (`<CN>`)
  * Use [`validation.cnf`](https://github.com/codice/ddf/files/2993047/validation.cnf.zip)
  * `openssl ca -keyfile private/cakey.pem -cert cacert.pem -in ocspSigning.csr -out ocspSigning.crt -config validation.cnf`
  * `openssl ocsp -index index.txt -port 8080 -rsigner ocspSigning.crt -rkey ocspSigning.key -CA cacert.pem -text -out log.txt`
  * Get the certificate from the p12 file. You can do
     * `openssl pkcs12 -in <hostName>.p12 -out certificate.crt` or
     * From Keystore Explorer, right click on the created keypair and select `Export > Export Certificate Chain`
  * In a new terminal window run in demoCA:
    * `openssl ocsp -CAfile cacert.pem -issuer cacert.pem -cert certificate.crt -url http://127.0.0.1:8080 -resp_text -noverify`

4. Add the OCSP server to DDF
  * Navigate to `System > Configuration > Online Certificate Status Protocol (OCSP)`
  * Add `http://127.0.0.1:8080` under `OCSP server URL`
  * `Enable validating a certificate against an OCSP server` option

5. Access DDF using the cert added to Chrome (may I suggest incognito mode)
  * When accessing DDF, Chrome will prompt you to choose a cert, select the one recently added
  * From DDF's login page, select `Sign in with certificate`
  * **You should be able to login**

6. Revoke the certificate
  * openssl ca -keyfile private/cakey.pem -cert cacert.pem -revoke certificate.crt
  * Restart the OCSP server
    * `openssl ocsp -index index.txt -port 8080 -rsigner ocspSigning.crt -rkey ocspSigning.key -CA cacert.pem -text -out log.txt`
   
7. Try signing in with the revoked cerificate  
  * Repeat step 6
  * **You should NOT be able to login**
  * **You should see a `Credentials are invalid` error**

8. Verify error alert
  * Enter an incorrect URL and verify that an error is posted to the admin console

#### What are the relevant tickets?
For GH Issues:
Fixes: #4469 

#### Screenshots
**Alert**
![screen shot 2019-03-06 at 12 18 16 pm](https://user-images.githubusercontent.com/29550062/53907272-0d63b680-400a-11e9-8bf8-bdc0d1ab409e.png)

**Documentation**
![screen shot 2019-02-28 at 5 17 41 pm](https://user-images.githubusercontent.com/29550062/53607634-d2323500-3b7c-11e9-8f6a-796e8ffc4251.png)

#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
